src/App.tsx

```diff
@@ -1,26 +1,204 @@
+  const [loadingTodoIds, setLoadingTodoIds] = useState<number[]>([]);
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then(setTodos)
+      .catch(() => setError(Error.UnableToLoadAll));
+  }, []);
+
+  const filteredTodos = useMemo(() => {
+    return filterTodos(todos, filter);
+  }, [todos, filter, filterTodos]);
+
+  const activeTodosCount = useMemo(() => {
+    return todos.filter(todo => !todo.completed).length;
+  }, [todos]);
+
+  const isCompletedTodos = useMemo(() => {
+    return todos.filter(todo => todo.completed).length > 0;
+  }, [todos]);
+
+  const addNewTodo = useCallback((title: string) => {
+    if (!title.trim()) {
+      setError(Error.NoTitle);
+
+      return;
+    }
+
+    const temp = {
+      title,
+      id: 0,
+      completed: false,
+      userId: USER_ID,
+    };
+
+    setTempTodo(temp);
+    setError(null);
+
+    addTodos(temp)
+      .then(res => {
+        setTodos(prev => [...prev, res]);
+      })
+      .catch(() => setError(Error.UnableToAdd))
+      .finally(() => {
+        setTempTodo(null);
+      });
+  }, []);
+
+  const deleteCurrentTodo = useCallback((id: number) => {
+    setLoadingTodoIds(prev => [...prev, id]);
+    deleteTodos(id)
+      .then(() => {
+        setTodos(prev => prev.filter(todo => todo.id !== id));
+      })
+      .catch(() => setError(Error.UnableToDelete))
+      .finally(() => {
+        setLoadingTodoIds([]);
+      });
+  }, []);
+
+  const deleteCompletedTodos = useCallback(() => {
+    const completedTodoIds = todos
+      .filter(todo => todo.completed)
+      .map(todo => todo.id);
+
+    setLoadingTodoIds(completedTodoIds);
+
+    Promise.all(completedTodoIds.map(id => deleteTodos(id)))
+      .then(() => {
+        setTodos(prev =>
+          prev.filter(todo => !completedTodoIds.includes(todo.id)),
+        );
+      })
+      .catch(() => setError(Error.UnableToDelete))
+      .finally(() => {
+        setLoadingTodoIds([]);
+      });
+  }, [todos]);
+
+  const toggleAllTodos = () => {
+    const areAllCompleted = todos.every(todo => todo.completed);
+
+    const updatedTodos = todos.map(todo => {
+      if (
+        // eslint-disable-next-line operator-linebreak
+        (areAllCompleted && !todo.completed) ||
+        (!areAllCompleted && todo.completed)
+      ) {
+        return {
+          ...todo,
+          completed: !todo.completed,
+        };
+      }
+
+      return todo;
+    });
+
+    const todosToUpdateIds = updatedTodos
+      .filter((todo, index) => todo.completed !== todos[index].completed)
+      .map(todo => todo.id);
+
+    setLoadingTodoIds(todosToUpdateIds);
+
+    const patchPromises = todosToUpdateIds.map(id =>
+      patchTodo(id, {
+        completed: updatedTodos.find(todo => todo.id === id)?.completed,
+      }),
+    );
+
+    Promise.all(patchPromises)
+      .then(() => {
+        setTodos(updatedTodos);
+      })
+      .catch(() => setError(Error.UnableToUpdate))
+      .finally(() => setLoadingTodoIds([]));
+  };
```

Comment:
This logic is not working properly. you need to get back for previous solution. You just need to create array of promises from uncompleted todos instead of making it for all of them.